### ISSUE TITLE
mime/multipart: Reader returns part that contains boundary string

### ISSUE THREAD

**Issue by tleyden** [golang/go#14085]
I'm seeing unexpected behavior with the Go HTTP MultipartReader when doing `multipart/related PUT` requests to a REST api which contain 1 JSON part and 3 image parts.  When calling `reader.NextPart()`, it seems to be returning data for both the 3rd and the 4th parts rather than just the 3rd part.  After digging into it further, I noticed that the data it returns for part 3 contains the boundary string itself, which definitely seems like it's breaking the parsing.

I've only seen this happen rarely, so I think it's dependent on the actual content of the data.

* Go version: `go1.5.1 darwin/amd64` 
* OS/Arch: Centos 7 Linux 64-bit + OSX 64-bit

Steps to reproduce:

* Run this [unit test](https://gist.github.com/tleyden/67ae66102dc43ee04690#file-testreadmultipartbody-go)
* Actual: test fails because it finds a part that contains a boundary string
* Expected: no part should returned from `reader.NextPart()` contain the boundary string  
* Expected: the multipart reader should be seeing 4 parts, but it's only seeing 3.  Should match up with what was [seen in wireshark](https://gist.github.com/tleyden/64def2e6c7e668247fec) for the request.

A hypothesis:

I did notice that this particular image data happens to contain the byte sequence for `\r\n`:

<img width="1421" alt="hexdump" src="https://cloud.githubusercontent.com/assets/296876/12538487/6e6bf01a-c28f-11e5-9e26-b71cc01bc30f.png">

Could that be causing [this line in multipart.go](https://github.com/golang/go/blob/master/src/mime/multipart/multipart.go#L236) to incorrectly segment it?

```
line, err := r.bufReader.ReadSlice('\n')
```

I'd be more than happy to try to distill this down to a repeatable unit test with no external dependencies.

**Comment by bradfitz** [comment]
This was fixed in Go 1.5.2: https://golang.org/doc/devel/release.html#go1.5.minor

(or at least, an identical-sounding bug was fixed. If you still see it with Go 1.5.2 or Go 1.5.3, let us know)

**Comment by tleyden** [comment]
Thanks!  I'll retest it.

**Comment by tleyden** [comment]
@bradfitz upgrading from 1.5.1 to 1.5.3 didn't change the outcome of the [unit test](https://gist.github.com/tleyden/67ae66102dc43ee04690#file-testreadmultipartbody-go) -- the go multipart module still can't correctly parse that particular multipart request.   

It might be caused by some subtlety about the way the original ObjectiveC-based client is sending the request, because in a [second unit test](https://gist.github.com/tleyden/41284837088e4f09fb15#file-testreadmultipartbody2) I tried using a go http client to reproduce the issue by sending a `multipart/related` request with approximately the same content, and couldn't reproduce any issues in either go 1.5.1 or 1.5.3. 

By "approximately the same content", I mean:

* The json part wasn't gzipped as it was in the original problematic request
* The boundary was different, `-9PWGbT-1_r1QpgvW0I-4E3` in failing vs `1174c809f7f3b164d57df5f5f23c652994408f51d135352b26e26a96ff6e ` in passing
* The image attachment parts were identical and in the same order

**Comment by bradfitz** [comment]
@adg, you have time to look at this? You'll be working a bit before me, otherwise I can look (my) Monday.

**Comment by tleyden** [comment]
@bradfitz @adg I managed to reproduce it via the go http client in a [third unit test](https://gist.github.com/tleyden/d7d66e06f82e79ec95dc#file-testreadmultipartbody3) -- the only difference from the second unit test is that it explicitly [sets the boundary string](https://gist.github.com/tleyden/d7d66e06f82e79ec95dc#file-testreadmultipartbody3-L67) to `-9PWGbT-1_r1QpgvW0I-4E3` in the client side mulitpart writer.

[Test output](https://gist.github.com/tleyden/d7d66e06f82e79ec95dc#file-test-output)

**Comment by bradfitz** [comment]
@tleyden, this bug report is already way above normal quality-wise, but if I could be greedy: can you whittle it down to a single file with no external data dependencies, suitable for writing a TestFoo(t *testing.T) function out of? Or even something that would work in play.golang.org? That's 95% of the work of fixing bugs, usually, writing a reproducible small test case.

**Comment by tleyden** [comment]
Yeah np, I'll give it a shot and post to the issue if I can come up with something.

**Comment by tleyden** [comment]
OK I managed to remove the external file dependencies and replaced them with zero'd out byte slices, and it's still reproducing the issue.  Admittedly it's still a bit clunky since it starts an http listener and goes through a client round trip.

* [TestReadMultipartBody4](https://gist.github.com/tleyden/c50334e3f3750911403d#file-testreadmultipartbody4)
* [Test output](https://gist.github.com/tleyden/c50334e3f3750911403d#file-test-results)

**Comment by adg** [comment]
@tleyden I tried running your TestReadMultipartBody4 against Go commit 801bebefa91205b0b69f2458701aac8169294884 and the test passed correctly. What version are you testing against?

```
=== RUN   TestReadMultipartBody4
2016/01/25 14:45:48 Writer using boundary: -9PWGbT-1_r1QpgvW0I-4E3
2016/01/25 14:45:48 attrs: map[boundary:-9PWGbT-1_r1QpgvW0I-4E3]
2016/01/25 14:45:48 Part: &{map[Content-Type:[application/json]] 0xc8200c82a0 0xc820948180 0  map[] {0xc8209a81e0}}
2016/01/25 14:45:48 Part data length: 2
2016/01/25 14:45:48 Part: &{map[Content-Disposition:[attachment; filename="result_image".]] 0xc8200f4070 0xc820948180 0  map[] {0xc820010230}}
2016/01/25 14:45:48 Part data length: 68881
2016/01/25 14:45:48 Part: &{map[Content-Disposition:[attachment; filename="source_image".]] 0xc8200f40e0 0xc820948180 0  map[] {0xc8200102d0}}
2016/01/25 14:45:48 Part data length: 97655
2016/01/25 14:45:48 Part: &{map[Content-Disposition:[attachment; filename="style_image".]] 0xc8200f4150 0xc820948180 0  map[] {0xc8200105f0}}
2016/01/25 14:45:48 Part data length: 85053
2016/01/25 14:45:48 err getting nextpart: EOF
2016/01/25 14:45:48 response status code: 200
--- PASS: TestReadMultipartBody4 (0.00s)
```

**Comment by tleyden** [comment]
I'm on go 1.5.3 on OSX:

```
$ go version
go version go1.5.3 darwin/amd64
```

I downloaded it from [here](https://storage.googleapis.com/golang/go1.5.3.src.tar.gz) and built it from source using `make.bash` on top of my previous go version (1.5.1)

**Comment by adg** [comment]
I'm able to reproduce the issue at go1.5.3, but the test passes against HEAD (~go1.6).

The question is whether it is worth fixing and releasing Go 1.5.4, or whether we should just recommend upgrading to go1.6, which should be out in a couple of weeks.

**Comment by bradfitz** [comment]
I don't remember fixes to this code this cycle. Did you narrow down which commit is responsible?

**Comment by adg** [comment]
I'll try bisecting.

**Comment by adg** [comment]
Looks like the fix happened in commit 821b54921a3cba5d853b531d4b03527c01bfc9b4

**Comment by adg** [comment]
Since this is fixed at HEAD, and Go 1.6 will be released soon, I'll close this issue.

**Comment by tleyden** [comment]
Works for me.  Thanks for getting to the bottom of this!
